### ISSUE TITLE
how to implement early stopping regularization?

### ISSUE THREAD

**Issue by tdhock** [mlr-org/mlr3torch#364]
Hi @sebffischer 

I am following this tutorial, https://mlr-org.github.io/mlr3torch-course/notebooks/6-mlr3torch.html
where it is written "we would not optimizer the number of epochs using offline tuning as shown here, but instead using early stopping, which is significantly more efficient and will be covered later."

I would like to define a Learner, that I can plug into benchmark(), with the number of epochs chosen to optimize a certain metric on the validation set (max AUC or min classification error).
In particular I would like to run the training until a large number of epochs, say 1000, then compute validation error/AUC for each epoch (and save these numbers for visualization/debugging/diagnostic purposes), and then figure out which is the best number of epochs, then re-initialize the model parameters and run the learning for the best number of epochs.
I know how to do this with regular torch, but I struggle to understand how to do it in the mlr3torch framework. Can you please help?

Here is what I tried. Following the tutorial, I ran this code,
```r
N.pixels <- 10
N.classes <- 2
N.features <- 100
N.images <- 200
set.seed(1)
my.X.mat <- matrix(runif(N.features*N.images), N.images, N.features)
my.df <- data.frame(y=factor(1:N.classes), my.X.mat)
my.task <- mlr3::TaskClassif$new("MyTask", my.df, target="y")
measure_list <- mlr3::msrs(c("classif.logloss", "classif.ce", "classif.auc"))
lrn_mlp <- mlr3::lrn("classif.mlp",
  neurons = c(50, 50),
  batch_size = 256,
  epochs = 30, # Number of training epochs
  device = "auto", # Uses GPU if available, otherwise CPU
  shuffle = TRUE, # because iris is sorted
  optimizer = mlr3torch::t_opt("adamw"),
  loss = mlr3torch::t_loss("cross_entropy"))
lrn_mlp$configure(
  validate = 0.3,
  callbacks = mlr3torch::t_clbk("history"),
  predict_type = "prob",
  measures_valid = measure_list,
  measures_train = measure_list)
lrn_mlp$train(my.task)
lrn_mlp$model$callbacks$history
```
The code above works for me, and gives me the table below of train/validation metrics for each epoch:
```r
> lrn_mlp$model$callbacks$history
Key: <epoch>
    epoch train.classif.logloss train.classif.ce train.classif.auc
    <num>                 <num>            <num>             <num>
 1:     1             0.6956844        0.5071429         0.5303123
 2:     2             0.6987848        0.5142857         0.4800980
 3:     3             0.6976289        0.4928571         0.4811186
 4:     4             0.6929046        0.4928571         0.5237804
 5:     5             0.6961695        0.5071429         0.4943866
 6:     6             0.6884315        0.4928571         0.5744029
 7:     7             0.6997247        0.5000000         0.4449888
 8:     8             0.6950990        0.5214286         0.4776485
 9:     9             0.6929091        0.4571429         0.5201061
10:    10             0.6949105        0.5000000         0.5039804
11:    11             0.7010106        0.6285714         0.4023270
12:    12             0.6962467        0.4928571         0.4909165
13:    13             0.6881956        0.4928571         0.5433762
14:    14             0.6909996        0.5071429         0.5264340
15:    15             0.6939516        0.5071429         0.4907124
16:    16             0.6963988        0.5142857         0.4831598
17:    17             0.6866738        0.4428571         0.5815473
18:    18             0.6851945        0.4357143         0.6044091
19:    19             0.6843430        0.3357143         0.6405389
20:    20             0.6869311        0.4357143         0.5964483
21:    21             0.6855669        0.4428571         0.5970606
22:    22             0.6939570        0.5500000         0.4627475
23:    23             0.6926428        0.4857143         0.5303123
24:    24             0.6933420        0.5000000         0.5178608
25:    25             0.6880914        0.4714286         0.5609308
26:    26             0.6872532        0.4642857         0.5719535
27:    27             0.6903958        0.4571429         0.5411308
28:    28             0.6850726        0.4000000         0.6133905
29:    29             0.6848338        0.4071429         0.5901204
30:    30             0.6828892        0.4000000         0.6133905
    epoch train.classif.logloss train.classif.ce train.classif.auc
    valid.classif.logloss valid.classif.ce valid.classif.auc
                    <num>            <num>             <num>
 1:             0.6966782        0.5166667         0.5528365
 2:             0.6955234        0.5166667         0.5850945
 3:             0.6947026        0.5166667         0.5828699
 4:             0.6941887        0.5166667         0.5828699
 5:             0.6935981        0.5166667         0.5951057
 6:             0.6930936        0.5333333         0.5951057
 7:             0.6926527        0.5333333         0.5761958
 8:             0.6922361        0.5166667         0.5717464
 9:             0.6919947        0.4333333         0.5661846
10:             0.6918702        0.4333333         0.5661846
11:             0.6917318        0.4666667         0.5583982
12:             0.6916043        0.4666667         0.5572859
13:             0.6915119        0.4333333         0.5572859
14:             0.6914860        0.4500000         0.5517241
15:             0.6915121        0.4833333         0.5483871
16:             0.6916126        0.4833333         0.5472747
17:             0.6916882        0.4833333         0.5450501
18:             0.6917182        0.4833333         0.5472747
19:             0.6917591        0.4833333         0.5383760
20:             0.6918070        0.5000000         0.5350389
21:             0.6919528        0.4500000         0.5272525
22:             0.6920563        0.4833333         0.5305895
23:             0.6921728        0.4666667         0.5228031
24:             0.6922410        0.4666667         0.5239155
25:             0.6923166        0.4666667         0.5272525
26:             0.6924423        0.4500000         0.5239155
27:             0.6925920        0.4500000         0.5239155
28:             0.6928159        0.4500000         0.5127920
29:             0.6929802        0.4500000         0.5094549
30:             0.6932022        0.4333333         0.5038932
    valid.classif.logloss valid.classif.ce valid.classif.auc
```
My question is how do I create a Learner that does this as an internal step of model training, automatically choosing the best number of epochs to optimize one of these?

I would like to present these results in a tutorial in April, so it would be great if I could get a response by next week.
Thanks in advance!

**Comment by sebffischer** [comment]
Hi @tdhock, we have a general mechanism for this that is covered in the mlr3 book: https://mlr3book.mlr-org.com/chapters/chapter15/predsets_valid_inttune.html.

In `mlr3torch`, you can enable early stopping by setting the `patience` parameter. It will use the first metric passed to `measures_valid` as the early stopping criterion. Afterwards, you can access the internally optimized values using `learner$internal_tuned_values`.

**Comment by tdhock** [comment]
Thanks for the quick response!
I was aware of the patience parameter, and that is similar to what I had in mind, but not exactly the same thing. (please correct me if I am wrong)
With patience parameter, it keeps going until some metric does not improve for a certain number of epochs, and then it stops. (and there is no retraining with an optimal number of epochs)
However I would like to do something different: keep going until a large number of epochs (say 1000), even if there is no improvement, we first always go up to the max number of epochs. Then we look at the subtrain/validation error curves and retrain with the optimal number of epochs. I prefer this method for teaching because it allows the students to clearly see the overfitting in the validation curve. Is there some easy way to implement this?

**Comment by sebffischer** [comment]
You could just do it manually like this. With the `patience = 100L` and `epochs = 100L` you essentially don't really early stop but will till compute the optimal epochs `$internal_tuned_values`.

Otherwise you could also just compute them manually from the history.

``` r
library(mlr3torch)
#> Loading required package: mlr3
#> Loading required package: mlr3pipelines
#> Loading required package: torch
library(mlr3tuning)
#> Loading required package: paradox

l = lrn("classif.mlp",
  epochs = 100L,
  measures_valid = msr("classif.logloss"),
  optimizer = t_opt("adamw", lr = 0.3),
  callbacks = t_clbk("history"),
  batch_size = 64,
  validate = 0.3,
  predict_type = "prob",
  patience = 100L
)

task = tsk("sonar")

l$train(task)

which.min(l$model$callbacks$history$valid.classif.logloss)
#> [1] 9
l$internal_tuned_values
#> $epochs
#> [1] 9

l$configure(
  epochs = l$internal_tuned_values$epochs,
  measures_valid = NULL,
  callbacks = NULL,
  validate = NULL,
  patience = 0
)

l$train(task)
```

<sup>Created on 2025-03-21 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

You can also automate both in a single learner, this is a little more complicated and requires an autotuner:

``` r
library(mlr3torch)
#> Loading required package: mlr3
#> Loading required package: mlr3pipelines
#> Loading required package: torch
library(mlr3tuning)
#> Loading required package: paradox

l = lrn("classif.mlp",
  epochs = to_tune(upper = 100L, internal = TRUE),
  measures_valid = msr("classif.logloss"),
  optimizer = t_opt("adamw", lr = 0.3),
  callbacks = t_clbk("history"),
  batch_size = 64,
  validate = 0.3,
  predict_type = "prob",
  patience = 100L
)

task = tsk("sonar")

# we don't need to make explicit predictions as validation
# is done anyway during training
# not neccessary, but saves some time
l$predict_sets = NULL

at = auto_tuner(
  learner = l,
  tuner = tnr("internal"),
  # the train/valid split is handled by the learner itself
  resampling = rsmp("insample"),
  # for the optimal model we will use the internal validation score
  # computed during training
  measure = msr("internal_valid_score", minimize = TRUE),
  # early stopping just needs a single run
  term_evals = 1,
  # so we can access the history afterwards
  store_models = TRUE
)

at$train(task)
#> INFO  [09:36:33.141] [bbotk] Starting to optimize 0 parameter(s) with '<TunerBatchInternal>' and '<TerminatorEvals> [n_evals=1, k=0]'
#> INFO  [09:36:33.156] [bbotk] Evaluating 1 configuration(s)
#> INFO  [09:36:33.164] [mlr3] Running benchmark with 1 resampling iterations
#> INFO  [09:36:33.192] [mlr3] Applying learner 'classif.mlp' on task 'sonar' (iter 1/1)
#> INFO  [09:36:34.139] [mlr3] Finished benchmark
#> INFO  [09:36:34.163] [bbotk] Result of batch 1:
#> INFO  [09:36:34.165] [bbotk]  internal_valid_score warnings errors runtime_learners internal_tuned_values
#> INFO  [09:36:34.165] [bbotk]             0.7628624        0      0            0.929             <list[1]>
#> INFO  [09:36:34.165] [bbotk]                                 uhash
#> INFO  [09:36:34.165] [bbotk]  8ef8fd3d-0a1a-4b36-b09d-e721f93a2f15
#> INFO  [09:36:34.181] [bbotk] Finished optimizing after 1 evaluation(s)
#> INFO  [09:36:34.181] [bbotk] Result:
#> INFO  [09:36:34.182] [bbotk]  internal_tuned_values learner_param_vals  x_domain internal_valid_score
#> INFO  [09:36:34.182] [bbotk]                 <list>             <list>    <list>                <num>
#> INFO  [09:36:34.182] [bbotk]              <list[1]>         <list[19]> <list[0]>            0.7628624
which.min(at$archive$learners(1)[[1L]]$model$callbacks$history$valid.classif.logloss)
#> [1] 10
at$tuning_result$internal_tuned_values
#> [[1]]
#> [[1]]$epochs
#> [1] 10
```

<sup>Created on 2025-03-21 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by tdhock** [comment]
Thanks! I did know about autotuner, but I did not know about tnr("internal") and msr("internal_valid_score"), that makes sense.

**Comment by tdhock** [comment]
Also, about the idea to use a patience parameter instead, I get another error:
```r
> glearner_linear_ce$configure(
+ validate = 0.5,
+ predict_type = "prob",
+ torch_model_classif.patience=10,
+ torch_model_classif.measures_valid = measure_list,
+ torch_model_classif.measures_train = measure_list)
> glearner_linear_ce$train(my.task)
WARN  [09:52:09.613] GraphLearner 'select.torch_ingress_num.nn_head.torch_loss.torch_optimizer.torch_model_classif' specifies a validation set, but none of its PipeOps use it.
Erreur : Learner 'classif.model' has measures_valid set, but its validate field is NULL`
This happened PipeOp torch_model_classif's $train()
```
The error message says the validate field is NULL, but I set validate=0.5 in configure(). Can you please help?
I also tried prefixing validate with torch_model_classif, but that did not work either:
```r
> glearner_linear_ce$configure(
+ torch_model_classif.validate = 0.5,
+ predict_type = "prob",
+ torch_model_classif.patience=10,
+ torch_model_classif.measures_valid = measure_list,
+ torch_model_classif.measures_train = measure_list)
Erreur : Cannot set argument 'torch_model_classif.validate' for 'GraphLearner' (not a parameter, not a field). Did you mean 'torch_model_classif.collate_fn' / 'torch_model_classif.eval_freq' / 'torch_model_classif.patience'?
```

**Comment by sebffischer** [comment]
You need to use `set_validate(glearner_linear_ce, validate = 0.5)`.

The problem with validation in the graph learner is that you need to set it on two levels:

1. Whether the graph as a whole does validation.
2. Which of the learners should use the validation data (there can be more than one).

The `set_validate()` convenience function tries to do the right thing and will work in your case.

**Comment by sebffischer** [comment]
I will try to get better error messages here, it really seems to be hard to configure all of this ...

**Comment by tdhock** [comment]
I was able to reproduce the auto-tuner result with MLP, thanks!

**Comment by tdhock** [comment]
also I confirm that set_validate() fixed the error.

There seems to be a few issues related to using the mlr3pipelines for defining a torch learner. I wonder if there is some simpler way to define a torch learner? (without using mlr3pipelines, something like lrn("mlp") but with more flexibility to define the architecture using convolutions etc)

**Comment by sebffischer** [comment]
> also I confirm that set_validate() fixed the error.
> 
> There seems to be a few issues related to using the mlr3pipelines for defining a torch learner. I wonder if there is some simpler way to define a torch learner? (without using mlr3pipelines, something like lrn("mlp") but with more flexibility to define the architecture using convolutions etc)

I think something like the following would be nice, is this what you have in mind?

```r
nn_custom = nn_module("nn_custom",
  initialize = function(...) {...},
 forward = function(...) {...}
)

learner = lrn("classif.module", module = nn_custom)
```

**Comment by tdhock** [comment]
yes something like that would be probably be easier to understand/use than the current pipeline API (at least for me). however we would need to separate the prediction and loss computation functions.

**Comment by sebffischer** [comment]
> yes something like that would be probably be easier to understand/use than the current pipeline API (at least for me). however we would need to separate the prediction and loss computation functions.

What do you mean with the second sentence?

**Comment by tdhock** [comment]
I mean that the interface should allow the user to define two modules:
* one that inputs the features and outputs model predictions (so user can use trained model for predictions on new data)
* one that inputs the model predictions and outputs the loss (for computing gradients during training)

right?
```r
lrn("classif.module", pred_module = pred_module, loss_module=torch::nn_mse_loss)
```
an interface like this via torch modules should be much easier to learn for somebody who knows about torch already, but does not know all the details of mlr3pipelines.

**Comment by sebffischer** [comment]
Thanks a lot for the feedback!

In `mlr3` we try to call `as_<object-type>` converters on arguments to make the API as flexible as possible.
Because `as_torch_loss` and `as_torch_optimizer` converters exist (e.g. convert `optim_adamw` to its `mlr3torch` counterpart), the following is already possible:

``` r
library(mlr3torch)
#> Loading required package: mlr3
#> Loading required package: mlr3pipelines
#> Loading required package: torch

graph = po("torch_ingress_num") %>>%
  po("nn_head") %>>%
  po("torch_loss", nn_cross_entropy_loss) %>>%
  po("torch_optimizer", optim_adamw) %>>%
  po("torch_model_classif")
```

<sup>Created on 2025-03-21 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by tdhock** [comment]
thanks for all your help here

**Comment by sebffischer** [comment]
FYI: https://github.com/mlr-org/mlr3torch/pull/371 makes it simple to create `mlr3torch` learners from an `nn_module_generator`
